### PR TITLE
Switch to using pgserver for postgres

### PR DESCRIPTION
**Pull request by orm011**
This enables pip install in mac to work without extra docker setup (unless they use nos)

- - -
This change is [Reviewable](https://reviewable.io/reviews/mkornacker/pixeltable/59)

### CODE REVIEW

File: pixeltable/env.py [hunk 1]
```diff
@@ -58,23 +56,14 @@ def __init__(self):
         # create logging handler to also log to stdout
         self._stdout_handler = logging.StreamHandler(stream=sys.stdout)
         self._stdout_handler.setFormatter(logging.Formatter(self._log_fmt_str))
+        self._inited = False
```
Comment (mkornacker) [inline pixeltable/env.py]: _initialized

File: pixeltable/env.py [hunk 2]
```diff
@@ -118,14 +111,12 @@ def set_up(self, echo: bool = False) -> None:
         self._log_dir = self._home / 'logs'
         self._tmp_dir = self._home / 'tmp'
 
+        env_pgdata = os.environ.get('PIXELTABLE_PGDATA')
```
Comment (mkornacker) [inline pixeltable/env.py]: it doesn't look like that's useful beyond accommodating tests, no?

File: pixeltable/env.py [hunk 3]
```diff
@@ -58,23 +56,14 @@ def __init__(self):
...
-        assert self._db_user is not None
-        assert self._db_password is not None
-        assert self._db_name is not None
-        assert self._db_port is not None
-        return (
-            f'postgresql://{self._db_user}:{"*****" if hide_passwd else self._db_password}'
-            f'@localhost:{self._db_port}/{self._db_name}'
-        )
-
+    def db_url(self, hide_passwd=False) -> str:
```
Comment (mkornacker) [inline pixeltable/env.py]: unused parameter

File: pixeltable/env.py [hunk 4]
```diff
@@ -159,11 +150,13 @@ def set_up(self, echo: bool = False) -> None:
         for path in glob.glob(f'{self._tmp_dir}/*'):
             os.remove(path)
 
-        # we now have a home directory; start runtime containers
-        self._set_up_runtime()
+        self._db_name = os.environ.get('PIXELTABLE_DB', 'pixeltable')
+        self.db_server = pgserver.get_server(self._pgdata_dir)
```
Comment (mkornacker) [inline pixeltable/env.py]: _...

File: pixeltable/tests/conftest.py [hunk 5]
```diff
@@ -15,19 +15,29 @@
...
+import tempfile
+import pathlib
 
 @pytest.fixture(scope='session')
 def init_env(tmp_path_factory) -> None:
     from pixeltable.env import Env
     # set the relevant env vars for Client() to connect to the test db
     home_dir = str(tmp_path_factory.mktemp('base') / '.pixeltable')
+
+    # NB: unix socket path names cannot exceed ~100 chars,
```
Comment (mkornacker) [inline pixeltable/tests/conftest.py]: remove 'NB' (if it's not worth noting, it's not worth having here as a comment; in general, don't add fluff to comments)

File: pixeltable/tests/conftest.py [hunk 6]
```diff
@@ -15,19 +15,29 @@
...
+
     os.environ['PIXELTABLE_HOME'] = home_dir
     test_db = 'test'
     os.environ['PIXELTABLE_DB'] = test_db
+    os.environ['PIXELTABLE_PGDATA'] = str(pgdata_dir)
+
     # this also runs create_all()
     Env.get().set_up(echo=True)
-    yield
+    yield Env.get()
```
Comment (mkornacker) [inline pixeltable/tests/conftest.py]: this changes the function signature

File: pixeltable/tests/conftest.py [hunk 7]
```diff
@@ -15,19 +15,29 @@
...
 
 @pytest.fixture(scope='session')
 def init_env(tmp_path_factory) -> None:
     from pixeltable.env import Env
     # set the relevant env vars for Client() to connect to the test db
     home_dir = str(tmp_path_factory.mktemp('base') / '.pixeltable')
+
+    # NB: unix socket path names cannot exceed ~100 chars, 
+    # but pgdata path from pytest fixture exceeds that length.
+    # Will make a separate dir at a shallower path for testing purposes.
```
Comment (mkornacker) [inline pixeltable/tests/conftest.py]: this seems like a general problem. is there no good solution for that? the user may well have a pixeltable home directory that will exceed this threshold.

**Review by mkornacker — CHANGES_REQUESTED** [review]

**Review by orm011 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mkornacker/pixeltable/59)* status: 0 of 4 files reviewed, 7 unresolved discussions (waiting on @mkornacker)

___
*[`pixeltable/env.py` line 114 at r1](https://reviewable.io/reviews/mkornacker/pixeltable/59#gh-1446542515:-NnjvoBsEJn7wriwDbU5:bso9afd) ([raw file](https://github.com/mkornacker/pixeltable/blob/5a1a7d84ecacb85632c60af8f45d4c756c4cf33b/pixeltable/env.py#L114)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

it doesn't look like that's useful beyond accommodating tests, no?
</blockquote></details>

correct.
___
*[`pixeltable/env.py` line 154 at r1](https://reviewable.io/reviews/mkornacker/pixeltable/59#gh-1446544336:-NnjyKwV8siyOtj2uWxU:b-wudljx) ([raw file](https://github.com/mkornacker/pixeltable/blob/5a1a7d84ecacb85632c60af8f45d4c756c4cf33b/pixeltable/env.py#L154)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

_...
</blockquote></details>

not sure what you mean here.
___
*[`pixeltable/tests/conftest.py` line 27 at r1](https://reviewable.io/reviews/mkornacker/pixeltable/59#gh-1446545960:-Nnjumq_34hewqnb5tS5:b-896fix) ([raw file](https://github.com/mkornacker/pixeltable/blob/5a1a7d84ecacb85632c60af8f45d4c756c4cf33b/pixeltable/tests/conftest.py#L27)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

remove 'NB' (if it's not worth noting, it's not worth having here as a comment; in general, don't add fluff to comments)
</blockquote></details>

Done.
___
*[`pixeltable/tests/conftest.py` line 29 at r1](https://reviewable.io/reviews/mkornacker/pixeltable/59#gh-1446550899:-NnjunHoAG0K3wOrJM1c:b-ju6if4) ([raw file](https://github.com/mkornacker/pixeltable/blob/5a1a7d84ecacb85632c60af8f45d4c756c4cf33b/pixeltable/tests/conftest.py#L29)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

this seems like a general problem. is there no good solution for that? the user may well have a pixeltable home directory that will exceed this threshold.
</blockquote></details>

There's definitely a solution, changing pgserver to decouple the socket path from the pgdata path.

The main issue with changing pgserver is that builds on github are slow. In the meantime, since the default setup is to use the $HOME/.pixeltable, which is unlikely to be an issue, seems ok to defer. 
___
*[`pixeltable/tests/conftest.py` line 40 at r1](https://reviewable.io/reviews/mkornacker/pixeltable/59#gh-1446546907:-Nnjvk8i4Xhmc5JJYLvd:b-igp9nk) ([raw file](https://github.com/mkornacker/pixeltable/blob/5a1a7d84ecacb85632c60af8f45d4c756c4cf33b/pixeltable/tests/conftest.py#L40)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

this changes the function signature
</blockquote></details>

removed, not needed in the end

**Review by orm011 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mkornacker/pixeltable/59)* status: 0 of 5 files reviewed, 7 unresolved discussions (waiting on @mkornacker)

___
*[`pixeltable/env.py` line 59 at r1](https://reviewable.io/reviews/mkornacker/pixeltable/59#gh-1446541655:-NnjyeuC6NbCvrw7fFVF:b-896fix) ([raw file](https://github.com/mkornacker/pixeltable/blob/5a1a7d84ecacb85632c60af8f45d4c756c4cf33b/pixeltable/env.py#L59)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

_initialized
</blockquote></details>

Done.
___
*[`pixeltable/env.py` line 61 at r1](https://reviewable.io/reviews/mkornacker/pixeltable/59#gh-1446542917:-NnjyfQ02J3bhcueF17R:bss4a2b) ([raw file](https://github.com/mkornacker/pixeltable/blob/5a1a7d84ecacb85632c60af8f45d4c756c4cf33b/pixeltable/env.py#L61)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

unused parameter
</blockquote></details>

Removed.

**Review by orm011 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mkornacker/pixeltable/59)* status: 0 of 5 files reviewed, 7 unresolved discussions (waiting on @mkornacker)

___
*[`pixeltable/env.py` line 114 at r1](https://reviewable.io/reviews/mkornacker/pixeltable/59#gh-1446542515:-NnlCaM-BeyHxWO4UhQz:b64hhbq) ([raw file](https://github.com/mkornacker/pixeltable/blob/5a1a7d84ecacb85632c60af8f45d4c756c4cf33b/pixeltable/env.py#L114)):*
<details><summary><i>Previously, orm011 (Oscar Moll) wrote…</i></summary><blockquote>

correct.
</blockquote></details>

no longer a thing, due to change in pgserver.
___
*[`pixeltable/tests/conftest.py` line 29 at r1](https://reviewable.io/reviews/mkornacker/pixeltable/59#gh-1446550899:-NnlCLE66w0VKA-i7j_c:b-k52p3d) ([raw file](https://github.com/mkornacker/pixeltable/blob/5a1a7d84ecacb85632c60af8f45d4c756c4cf33b/pixeltable/tests/conftest.py#L29)):*
<details><summary><i>Previously, orm011 (Oscar Moll) wrote…</i></summary><blockquote>

There's definitely a solution, changing pgserver to decouple the socket path from the pgdata path.

The main issue with changing pgserver is that builds on github are slow. In the meantime, since the default setup is to use the $HOME/.pixeltable, which is unlikely to be an issue, seems ok to defer.
</blockquote></details>

no longer relevant, handled in new pgserver by using a different path for socket.

File: pixeltable/env.py [hunk 8]
```diff
@@ -58,23 +56,14 @@ def __init__(self):
...
-        assert self._db_user is not None
-        assert self._db_password is not None
-        assert self._db_name is not None
-        assert self._db_port is not None
-        return (
-            f'postgresql://{self._db_user}:{"*****" if hide_passwd else self._db_password}'
-            f'@localhost:{self._db_port}/{self._db_name}'
-        )
-
+    def db_url(self) -> str:
```
Comment (mkornacker) [inline pixeltable/env.py:64]: the other members like tmp_dir are properties, might as well make this one

File: pixeltable/env.py [hunk 9]
```diff
@@ -58,23 +56,14 @@ def __init__(self):
...
+        return self._db_url
+    
     @property
     def db_service_url(self) -> str:
```
Comment (mkornacker) [inline pixeltable/env.py]: do we still need this?

File: pixeltable/env.py [hunk 10]
```diff
@@ -117,17 +110,13 @@ def set_up(self, echo: bool = False) -> None:
         self._dataset_cache_dir = self._home / 'dataset_cache'
         self._log_dir = self._home / 'logs'
         self._tmp_dir = self._home / 'tmp'
+        self._pgdata_dir = (self._home / 'pgdata')
```
Comment (mkornacker) [inline pixeltable/env.py:113]: only __init__ should create instance members

File: pixeltable/env.py [hunk 11]
```diff
@@ -175,40 +168,16 @@ def set_up(self, echo: bool = False) -> None:
...
                 conn.execute(sql.text('CREATE EXTENSION vector'))
-
         else:
-            self._logger.info(f'found database {self.db_url(hide_passwd=True)}')
+            self._logger.info(f'found database {self.db_url()}')
             if self._sa_engine is None:
                 self._sa_engine = sql.create_engine(self.db_url(), echo=echo, future=True)
             metadata.upgrade_md(self._sa_engine)
 
+        # we now have a home directory and db; start runtime containers
```
Comment (mkornacker) [inline pixeltable/env.py]: update comment

File: pixeltable/env.py [hunk 12]
```diff
@@ -253,46 +220,9 @@ def _set_up_runtime(self) -> None:
...
 
-    def _postgres_image(self) -> str:
-        if self._is_apple_cpu():
-            return 'ankane/pgvector:latest'
-            #return 'arm64v8/postgres:15-alpine'
-        else:
-            #return 'postgres:15-alpine'
-            return 'ankane/pgvector:latest'
-
-    def tear_down(self) -> None:
```
Comment (mkornacker) [inline pixeltable/env.py:292]: isn't this needed for the tests?

File: pixeltable/tests/conftest.py [hunk 13]
```diff
@@ -16,19 +16,19 @@
...
     home_dir = str(tmp_path_factory.mktemp('base') / '.pixeltable')
+
     os.environ['PIXELTABLE_HOME'] = home_dir
     test_db = 'test'
     os.environ['PIXELTABLE_DB'] = test_db
+
     # this also runs create_all()
     Env.get().set_up(echo=True)
     yield
-    # leave db in place for debugging purposes
```
Comment (mkornacker) [inline pixeltable/tests/conftest.py]: this comment is useful

**Review by orm011 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mkornacker/pixeltable/59)* status: 0 of 5 files reviewed, 13 unresolved discussions (waiting on @mkornacker)

___
*[`pixeltable/env.py` line 0 at r2](https://reviewable.io/reviews/mkornacker/pixeltable/59#gh-1446821441:-NnlFipoAw7OEdedxW0A:b-xpyly5) ([raw file](https://github.com/mkornacker/pixeltable/blob/259c69f4f40accea5ca05b72f01276e068e47576/pixeltable/env.py#L0)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

*(Reviewable was [unable](https://github.com/Reviewable/Reviewable/issues/14) to map [this GitHub inline comment thread](https://github.com/mkornacker/pixeltable/pull/59#discussion_r1446821441) to the right spot — sorry!)*

isn't this needed for the tests?
</blockquote></details>

it was only called from one place in the code base, I simply inlined that use case.
___
*[`pixeltable/env.py` line 61 at r2](https://reviewable.io/reviews/mkornacker/pixeltable/59#gh-1446819532:-NnlJ0fE-T6jdHgU8Hyr:b-896fix) ([raw file](https://github.com/mkornacker/pixeltable/blob/259c69f4f40accea5ca05b72f01276e068e47576/pixeltable/env.py#L61)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

the other members like tmp_dir are properties, might as well make this one
</blockquote></details>

Done.
___
*[`pixeltable/env.py` line 65 at r2](https://reviewable.io/reviews/mkornacker/pixeltable/59#gh-1446819745:-NnlJ1Yk4d5VEfXZVYLT:b-hfpv0l) ([raw file](https://github.com/mkornacker/pixeltable/blob/259c69f4f40accea5ca05b72f01276e068e47576/pixeltable/env.py#L65)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

do we still need this?
</blockquote></details>

removed
___
*[`pixeltable/env.py` line 113 at r2](https://reviewable.io/reviews/mkornacker/pixeltable/59#gh-1446820299:-NnlJ3-A1u2SPfglnXSc:bb7f2mf) ([raw file](https://github.com/mkornacker/pixeltable/blob/259c69f4f40accea5ca05b72f01276e068e47576/pixeltable/env.py#L113)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

only __init__ should create instance members
</blockquote></details>

done
___
*[`pixeltable/env.py` line 177 at r2](https://reviewable.io/reviews/mkornacker/pixeltable/59#gh-1446821114:-NnlJ5qQ4GfSk0HNqltQ:b-mj3fma) ([raw file](https://github.com/mkornacker/pixeltable/blob/259c69f4f40accea5ca05b72f01276e068e47576/pixeltable/env.py#L177)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

update comment
</blockquote></details>

isnt the comment still relevant and accurate? 
___
*[`pixeltable/tests/conftest.py` line 0 at r2](https://reviewable.io/reviews/mkornacker/pixeltable/59#gh-1446821664:-NnlJN4_Dn_2NfW2mmKu:boquah9) ([raw file](https://github.com/mkornacker/pixeltable/blob/259c69f4f40accea5ca05b72f01276e068e47576/pixeltable/tests/conftest.py#L0)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

*(Reviewable was [unable](https://github.com/Reviewable/Reviewable/issues/14) to map [this GitHub inline comment thread](https://github.com/mkornacker/pixeltable/pull/59#discussion_r1446821664) to the right spot — sorry!)*

this comment is useful
</blockquote></details>

restored.

**Review by orm011 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mkornacker/pixeltable/59)* status: 0 of 5 files reviewed, 13 unresolved discussions (waiting on @mkornacker)

___
*[`pixeltable/env.py` line 113 at r2](https://reviewable.io/reviews/mkornacker/pixeltable/59#gh-1446820299:-NnlKjfd2seyvZpKpWrl:b-fh7hc8) ([raw file](https://github.com/mkornacker/pixeltable/blob/259c69f4f40accea5ca05b72f01276e068e47576/pixeltable/env.py#L113)):*
<details><summary><i>Previously, orm011 (Oscar Moll) wrote…</i></summary><blockquote>

done
</blockquote></details>

ps. can we just set up ruff for linting? this kind of stuff makes no difference to customers, and just represented another iteration pushing and testing... and tests work either way. makes python dev much slower.

**Review by mkornacker — APPROVED** [review]

**Pull request closed without merging**